Math/NumberTheory/EisensteinIntegers.hs

```diff
@@ -0,0 +1,345 @@
+    negate (a :+ b) = (-a) :+ (-b)
+    fromInteger n = n :+ 0
+    signum z@(a :+ b)
+        | a == 0 && b == 0 = z               -- hole at origin
+        | otherwise        = z `divE` abs z
+
+-- | List of all Eisenstein units, counterclockwise across all sextants,
+-- starting with @1@.
+ids :: [EisensteinInteger]
+ids = take 6 (iterate ((1 + ω) *) 1)
+
+-- | Produce a list of an @EisensteinInteger@'s associates.
+associates :: EisensteinInteger -> [EisensteinInteger]
+associates e = map (e *) ids
+
+-- | Takes an Eisenstein prime whose norm is of the form @3k + 1@ with @k@
+-- a nonnegative integer.
+-- * Does *not* check for this precondition.
+-- * @head@ will fail when supplied a number unsatisfying it.
+primary :: EisensteinInteger -> EisensteinInteger
+primary = head . filter (\p -> p `modE` 3 == 2) . associates
+
+-- | Simultaneous 'quot' and 'rem'.
+quotRemE
+    :: EisensteinInteger
+    -> EisensteinInteger
+    -> (EisensteinInteger, EisensteinInteger)
+quotRemE = divHelper quot
+
+-- | Eisenstein integer division, truncating toward zero.
+quotE :: EisensteinInteger -> EisensteinInteger -> EisensteinInteger
+n `quotE` d = q where (q,_) = quotRemE n d
+
+-- | Eisenstein integer remainder, satisfying
+--
+-- > (x `quotE` y)*y + (x `remE` y) == x
+remE :: EisensteinInteger -> EisensteinInteger -> EisensteinInteger
+n `remE`  d = r where (_,r) = quotRemE n d
+
+-- | Simultaneous 'div' and 'mod' of Eisenstein integers.
+divModE
+    :: EisensteinInteger
+    -> EisensteinInteger
+    -> (EisensteinInteger, EisensteinInteger)
+divModE = divHelper div
+
+-- | Eisenstein integer division, truncating toward negative infinity.
+divE :: EisensteinInteger -> EisensteinInteger -> EisensteinInteger
+n `divE` d = q where (q,_) = divModE n d
+
+-- | Eisenstein integer remainder, satisfying
+--
+-- > (x `divE` y)*y + (x `modE` y) == x
+modE :: EisensteinInteger -> EisensteinInteger -> EisensteinInteger
+n `modE` d = r where (_,r) = divModE n d
+
+-- | Function that does most of the underlying work for @divMod@ and
+-- @quotRem@, apart from choosing the specific integer division algorithm.
+-- This is instead done by the calling function (either @divMod@ which uses
+-- @div@, or @quotRem@, which uses @quot@.)
+divHelper
+    :: (Integer -> Integer -> Integer)
+    -> EisensteinInteger
+    -> EisensteinInteger
+    -> (EisensteinInteger, EisensteinInteger)
+divHelper divide g h =
+    let nr :+ ni = g * conjugate h
+        denom = norm h
+        q = divide nr denom :+ divide ni denom
+        p = h * q
+    in (q, g - p)
+
+-- | Conjugate a Eisenstein integer.
+conjugate :: EisensteinInteger -> EisensteinInteger
+conjugate (a :+ b) = (a - b) :+ (-b)
+
+-- | The square of the magnitude of a Eisenstein integer.
+norm :: EisensteinInteger -> Integer
+norm (a :+ b) = a*a - a * b + b*b
+
+-- | Checks if a given @EisensteinInteger@ is prime. @EisensteinInteger@s
+-- whose norm is a prime congruent to @0@ or @1@ modulo 3 are prime.
```

Comment:
I suggest to write about norms modulo 6, because it is more specific.

Reply:
I don't think it is necessary. Counting with associates, there are 6 primes with norm 3, and 6 with norm 4. These norms are not congruent to 0 or 1 modulo 6.